pkg/metrics/metrics_grabber.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/master/ports"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/golang/glog"
+)
+
+type MetricsCollection struct {
+	ApiServerMetrics         ApiServerMetrics
+	ControllerManagerMetrics ControllerManagerMetrics
+	KubeletMetrics           map[string]KubeletMetrics
+	SchedulerMetrics         SchedulerMetrics
+}
+
+type MetricsGrabber struct {
+	client                    *client.Client
+	grabFromApiServer         bool
+	grabFromControllerManager bool
+	grabFromKubelets          bool
+	grabFromScheduler         bool
+	kubeletPort               int
+	masterName                string
+	registeredMaster          bool
+}
+
+func NewMetricsGrabber(c *client.Client, kubelets bool, scheduler bool, controllers bool, apiServer bool) (*MetricsGrabber, error) {
+	kubeletPort := 0
+	registeredMaster := false
+	masterName := ""
+	nodeList, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if len(nodeList.Items) < 1 {
+		return nil, fmt.Errorf("Can't find any Nodes in the API server to grab metrics from")
+	}
+	for _, node := range nodeList.Items {
+		if strings.HasSuffix(node.Name, "master") {
```

Comment:
This is a hacky solution. Can you please create a separate function like isMasterNode and leave a TODO there to make it less hacky :)